Add `FromArma` and `IntoArma` derive proc macros

Eathox opened the pull request:
Based on your WIP branch this PR adds derive macros. I decided to make two separate proc macros to allow for a custom implementation of one trait while also deriving the other.

The proc macros support deriving implementations for structs with named and unnamed fields only. I'm sure this feature can later be expanded to also support enums but I imagine this is a fine starting point.

I did still have some questions, that I'd like your input on:
- Currently there is no support for unit structs as I'm not sure what would be a good representation, maybe the structs name?
- ~~There's a discrepancy between how `struct EmptyNamed{}` and `struct EmptyUnnamed()` are handled as `EmptyUnnamed` is currently not supported while `EmptyNamed` expects and results in `"[]"`~~. `struct EmptyNamed{}` and `struct EmptyUnnamed()` are now treated as unit structs. Is this acceptable or would you like to have this handled in a different way?

Eathox commented:
Heads up that I've also added a new error type `FromArmaError`. I think it makes things cleaner and could come in handy if someone wants to write their own `FromArma` implementation. Let me know if you have any thoughts or if you think changes are needed, thanks

BrettMayson commented:
I think `#[arma(default)]` is better to have per field, like serde. Since there may be fields you don't want to allow to use a default, while others are fine

Eathox commented:
I'm on it. And additionally we can still keep the top level default, like serde (also has both).

Eathox commented:
Apologies for the the down time. I have added support for field attributes and implemented default. Additionally default can now also be used with tuple structs not just map-like ones. Lastly most error reporting is now on the attribute's span, so slightly nicer

BrettMayson commented:
Sorry, so is this good to go, assuming the resolved conflict?

Eathox commented:
No worries. Yep its ready to go, all thought I did just see I forgot to make readme changes.
If you want I can get around to that somewhere latter this week.

Eathox commented:
Sorry I've been rather busy lately. I've now added a short section to the README and opted to have the attribute documentation in the code it self to also have em easily available while using an IDE. I once again believe this is rdy for merge, please let me know if there's anything you'd like changed
